Add Put.io as new backend

cenkalti opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

To add put.io as a new backend.

#### Was the change discussed in an issue or in the forum before?

No. But, put.io is already supported through WebDAV interface as listed in rclone.org main page. The support was read-only though. This PR adds write support and makes it possible to use 2-way sync commands. BTW, I work at put.io.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

ncw commented:
Great job :-)

Do the integration tests all pass?

Will this work on a free account?  If not can you supply one for the integration test runner? I'm nick@craig-wood.com if you want to discuss this.

I'll give it a test then a review and we can squeeze this into 1.49 which is due at the weekend if it all checks out!

cenkalti commented:
Hi @ncw. All of the unit tests are passed.

Only one test in fs/operations fails: `TestCopyFileBackupDir`. This commit (https://github.com/cenkalti/rclone/commit/29db41aafab447fdbb0426adf589a2906968526e) fixes the test. I think the problem is the test, not the backend code. Because, when I run only this test, it passes. The command I use to run this test is:
`go test -v -remote TestPutio: -run TestCopyFileBackupDir`

However, when I run all of the tests, it fails. I have found that this test only fails if `TestMoveFileBackupDir` has run before in the same run. The command I use to reproduce:
`go test -verbose -v -remote TestPutio: -run 'BackupDir$'`

I also tried this with `dropbox` backend and it does not work either. I guess, there is a problem with the `fs/cache` used it tests. `Finalise` is called in each test to ensure all files in the remote are deleted, including the `BackupDir`. When the next test is run, this BackupDir fs is retrieved from the cache which is not valid anymore. Can you help me to fix the problem? Maybe there is problem with my `fs.Move` implementation.

Also, 2 tests in fs/sync fails: `TestSyncBackupDirWithSuffix` and `TestSyncBackupDirWithSuffixKeepExtension`. This is due to the backup dir is being removed when `Finalise` called in `TestSyncBackupDir`.

Please let me know if you have more information. That would be great if we can include this in 1.49.

Put.io does not offer free accounts but I will be sending you the credentials of a test account.

ncw commented:
> Hi @ncw. All of the unit tests are passed.
> 
> Only one test in fs/operations fails: `TestCopyFileBackupDir`. This commit ([cenkalti@29db41a](https://github.com/cenkalti/rclone/commit/29db41aafab447fdbb0426adf589a2906968526e)) fixes the test. I think the problem is the test, not the backend code. Because, when I run only this test, it passes. The command I use to run this test is:
> `go test -v -remote TestPutio: -run TestCopyFileBackupDir`
> 
> However, when I run all of the tests, it fails. I have found that this test only fails if `TestMoveFileBackupDir` has run before in the same run. The command I use to reproduce:
> `go test -verbose -v -remote TestPutio: -run 'BackupDir$'`
> 
> I also tried this with `dropbox` backend and it does not work either. I guess, there is a problem with the `fs/cache` used it tests. `Finalise` is called in each test to ensure all files in the remote are deleted, including the `BackupDir`. When the next test is run, this BackupDir fs is retrieved from the cache which is not valid anymore. Can you help me to fix the problem? Maybe there is problem with my `fs.Move` implementation.

Well spotted.  You worked out the underlying problem perfectly - the cache is a new feature.

I made a patch to clear the cache between all test runs which I think is the right thing to do.

The problem happened because of this
- we make a new backend as a subdirectory of the old one "remote:backup"
- we use this in the first test
- its cache becomes full of stuff
- we purge "remote:" but do not purge "remote:backup".  Since their directory caches are independent, "remote:backup" now contains stale entries.
- we then reuse "remote:backup" via the fs cache
- the stale entry causes the test to fail.

> Also, 2 tests in fs/sync fails: `TestSyncBackupDirWithSuffix` and `TestSyncBackupDirWithSuffixKeepExtension`. This is due to the backup dir is being removed when `Finalise` called in `TestSyncBackupDir`.

Hopefully the same fix should fix these.

> Please let me know if you have more information. That would be great if we can include this in 1.49.
>
> Put.io does not offer free accounts but I will be sending you the credentials of a test account.

Received thank you :-)

ncw commented:
> Hi @ncw. All of the unit tests are passed.
> 
> Only one test in fs/operations fails: `TestCopyFileBackupDir`. This commit ([cenkalti@29db41a](https://github.com/cenkalti/rclone/commit/29db41aafab447fdbb0426adf589a2906968526e)) fixes the test. I think the problem is the test, not the backend code. Because, when I run only this test, it passes. The command I use to run this test is:
> `go test -v -remote TestPutio: -run TestCopyFileBackupDir`
> 
> However, when I run all of the tests, it fails. I have found that this test only fails if `TestMoveFileBackupDir` has run before in the same run. The command I use to reproduce:
> `go test -verbose -v -remote TestPutio: -run 'BackupDir$'`
> 
> I also tried this with `dropbox` backend and it does not work either. I guess, there is a problem with the `fs/cache` used it tests. `Finalise` is called in each test to ensure all files in the remote are deleted, including the `BackupDir`. When the next test is run, this BackupDir fs is retrieved from the cache which is not valid anymore. Can you help me to fix the problem? Maybe there is problem with my `fs.Move` implementation.

Well spotted.  You worked out the underlying problem perfectly - the cache is a new feature.

I made a patch to clear the cache between all test runs which I think is the right thing to do.

The problem happened because of this
- we make a new backend as a subdirectory of the old one "remote:backup"
- we use this in the first test
- its cache becomes full of stuff
- we purge "remote:" but do not purge "remote:backup".  Since their directory caches are independent, "remote:backup" now contains stale entries.
- we then reuse "remote:backup" via the fs cache
- the stale entry causes the test to fail.

> Also, 2 tests in fs/sync fails: `TestSyncBackupDirWithSuffix` and `TestSyncBackupDirWithSuffixKeepExtension`. This is due to the backup dir is being removed when `Finalise` called in `TestSyncBackupDir`.

Hopefully the same fix should fix these.

> Please let me know if you have more information. That would be great if we can include this in 1.49.
>
> Put.io does not offer free accounts but I will be sending you the credentials of a test account.

Received thank you :-)

Can you rebase on master and check that the fix works for you?

ncw commented:
Can you rebase on master and try your tests again?

cenkalti commented:
All unit and integration tests have run successfully on my computer.

I have rebased my commit on master and updated the PR.

I have also extended the integration test account validity to 10 years.

ncw commented:
> All unit and integration tests have run successfully on my computer.

:-)
 
> I have rebased my commit on master and updated the PR.

Thanks

> I have also extended the integration test account validity to 10 years.

Much appreciated, thank you :-)

I've given the backend a bit of a test and it seems to work very well - well done :-)

My only comment is that it is a bit noisy when run with `-vv`.  Can you comment out the entry/exit `fs.Debugf` in the backend?  For the normal paths list/upload/download users will not expect debugs, maybe with the exception of multi part uploads.

Note for the future - the Trace function is really useful for this kind of debugging

https://github.com/rclone/rclone/blob/341d8800279c6264cd7a3f8fc539200fb51bd5ce/fs/log/log.go#L38

Can you also do a little bit of squashing of your commits?

I normally squash backend commits into 1 for the backend and 1 for the vendor dependencies.  You'll need 1 more for the CRC hash.

So these 3 should be one commit

```
commit 8e0d50ea0c2647fa52795070eb9943d39d1ef741 (HEAD -> pr-3469-putio)
Date:   Wed Aug 21 22:40:32 2019 +0300

    putio: fix updating modtime; escape time in query

commit cb389ee3257ddc2ffdf0a16c86d4524805f9afaf
Date:   Thu Aug 8 09:19:13 2019 +0300

    putio: add docs

commit be8a61ce4fe3c5eb8b1e73493eaff41b1409676a
Date:   Tue Aug 6 15:47:52 2019 +0300

    putio: add new backend
```

These should be one commit starting with `vendor: add github.com/x/y for putio client`

```
commit 3a7f9e37a196a45a2f3a59a29663d0989a53530b
Date:   Tue Aug 6 15:44:18 2019 +0300

    putio: vendor client package

commit 0a322858cc180289cfb6b40b6ecccc28f948d6a9
Date:   Tue Aug 6 15:43:14 2019 +0300

    putio: add client library dependency
```

And that is perfect as it is :-)

```
commit c4b642ed71988f71fd2995054aafffe94aa6b22e
Date:   Tue Aug 6 15:39:29 2019 +0300

    hash: add CRC-32 support
```

If you can make the backend a bit quieter and do a bit of squashing then I'll merge what you've done, we can get it into the integration tests and fixup anything else in master.

Looking very good - thank you :-)

cenkalti commented:
I have squashed the PR into 3 commits.

I have replaced Debugf calls with Trace calls, then commented them out.

I have only left the Debugf calls that are just before API calls to put.io. If you want, I can comment them out too.

ncw commented:
> I have squashed the PR into 3 commits.

Perfect
 
> I have replaced Debugf calls with Trace calls, then commented them out.

Great
 
> I have only left the Debugf calls that are just before API calls to put.io. If you want, I can comment them out too.

That is great thank you :-)

I think I would comment out nearly all the remaining Debugf~s  (comment them out rather than deleting them - we can uncomment them again if we really need them!  Note that most backends are easiest to debug with `-vv --dump bodies` which show the HTTP accesses).

```
backend/putio/fs.go:	fs.Debugf(p, "Root id: %s", p.dirCache.RootID())
backend/putio/fs.go:		fs.Debugf(f, "creating folder. part: %s, parentID: %d", leaf, parentID)
backend/putio/fs.go:		fs.Debugf(f, "listing file: %d", fileID)
backend/putio/fs.go:		fs.Debugf(f, "listing files inside List: %d", parentID)
backend/putio/fs.go:		// fs.Debugf(f, "child: %s", remote)
backend/putio/fs.go:		fs.Debugf(f, "getting file: %d", fileID)
LEAVE backend/putio/fs.go:			fs.Debugf(f, "Sending zero length chunk")
LEAVE backend/putio/fs.go:			fs.Debugf(f, "Sending chunk. start: %d length: %d", start, reqSize)
backend/putio/fs.go:		fs.Debugf(f, "listing files: %d", dirID)
backend/putio/fs.go:		fs.Debugf(f, "deleting file: %d", dirID)
backend/putio/fs.go:		fs.Debugf(f, "deleting file: %d", rootID)
backend/putio/fs.go:		fs.Debugf(f, "copying file (%d) to parent_id: %s", srcObj.file.ID, directoryID)
backend/putio/fs.go:		fs.Debugf(f, "moving file (%d) to parent_id: %s", srcObj.file.ID, directoryID)
backend/putio/fs.go:		fs.Debugf(f, "moving file (%s) to parent_id: %s", srcID, dstDirectoryID)
backend/putio/fs.go:		fs.Debugf(f, "getting account info")
backend/putio/fs.go:		fs.Debugf(f, "emptying trash")
backend/putio/object.go:		fs.Debugf(o, "requesting child. directoryID: %s, name: %s", directoryID, leaf)
LEAVE backend/putio/object.go:			fs.Debugf(o, "Failed to read metadata: %v", err)
backend/putio/object.go:	fs.Debugf(o, "setting modtime: %s", modTime.String())
backend/putio/object.go:		fs.Debugf(o, "opening file: id=%d", o.file.ID)
backend/putio/object.go:		fs.Debugf(o, "removing file: id=%d", o.file.ID)
```

cenkalti commented:
@ncw Done commenting the remaining Debugf calls.

ncw commented:
Perfect!  I'll merge this now :-)

Thank you very much for your excellent work - a very smooth pull request for a backend.

Look out for https://pub.rclone.org/integration-tests/current/ tomorrow morning to see whether the integration tests all passed on the integration test machine :crossed_fingers:

ncw commented:
Do you want to ask for testers in [the rclone forum](https://forum.rclone.org/)?  It will be in [the latest beta](https://beta.rclone.org) in 15-30 mins.

cenkalti commented:
I have posted an announcement in the forum: https://forum.rclone.org/t/put-io-looking-for-beta-testers/11400

@ncw Thank you.
